app/models/recipients.rb

```diff
@@ -17,18 +17,23 @@ def user_emails
 
   def users
     return @users if @users
-    @users = User.joins(:teams, :roles).where(roles: { name: roles })
-    if group.to_sym == :selected_teams
-      @users = @users.where(teams: { kind: Team::KINDS })
-    elsif group.to_sym == :unselected_teams
-      @users = @users.where(teams: { kind: nil })
-    end
-    if seasons.any?
-      @users = @users.where(teams: {
-        season: Season.where(name: seasons).pluck(:id)
-      })
-    end
-    @users.uniq
+    @users = User.joins(:roles).where(roles: { name: roles })
+    @users = @users.where('roles.team_id IN (?) OR roles.name IN (?)', teams, teamless_roles) if filter_by_teams?
+    @users = @users.uniq
+    @users
+  end
+
+  def filter_by_teams?
+    group.to_sym != :everyone || seasons.any?
+  end
+
+  def teams
+    return @teams if @teams
+    @teams = Team
+    @teams = @teams.where(kind: group.to_sym == :selected_teams ? Team::KINDS : nil) if group.to_sym != :everyone
```

Comment:
So this is a little tricky to understand, it took me a while too.

There are a number of filters, `group`, `seasons` and `to`.

If `group` is `:selected_teams` it will select any team that has a `kind` that is included in `Team::KINDS`, otherwise select where `kind` is specifically `nil`. `group: :everyone` will select every team.

It gets more tricky with [line 21](https://github.com/rails-girls-summer-of-code/rgsoc-teams/pull/298/files#diff-875df54ccb6e747fd165f1ebcf5d3266R21)

```ruby
@users = @users.where('roles.team_id IN (?) OR roles.name IN (?)', teams, teamless_roles) if filter_by_teams?
```

where if we need to filter by team it selects users who's role belongs to a team or users who's roles is teamless (organizer developer helpdesk) and has been included in the `to` filter.

I have considered rewriting the spec for `recipients.rb`, it sort of needs it because there are a lot of permutations that are not considered in the spec.